ttnn/cpp/ttnn/tensor/vector_base.hpp

```diff
@@ -0,0 +1,41 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+namespace tt::tt_metal {
+
+// Container wrapper that allows negative indexing
+class vector_base {
+public:
+    vector_base() = default;
+    explicit vector_base(const std::vector<uint32_t>& shape) : mValue(shape) {}
+    explicit vector_base(std::vector<uint32_t>&& shape) : mValue(std::move(shape)) {}
+    explicit vector_base(std::initializer_list<uint32_t> ilist) : mValue(ilist) {}
+    template<std::size_t N>
+    explicit vector_base(const std::array<uint32_t, N>& arr) : mValue(arr.begin(), arr.end()) {}
+
+    template<std::size_t N>
+    bool operator==(const std::array<uint32_t, N> &other) const {
+        bool sameSize = mValue.size() == N;
+        return sameSize && std::equal(mValue.begin(), mValue.end(), other.begin());
+    }
+
+    bool operator==(const vector_base &other) const;
+    bool operator==(const std::vector<uint32_t> &other) const;
+
+    uint32_t operator[](int32_t index) const;
+    uint32_t &operator[](int32_t index);
+
+
+    auto cbegin() const { return this->mValue.cbegin(); }
+    auto cend() const { return this->mValue.cend(); }
+
+    [[nodiscard]] const std::vector<uint32_t>& as_vector() const { return this->mValue; }
+
+protected:
+    std::vector<uint32_t> mValue;
```

Comment:
Lets rename mValue -> m_value 
vector_base -> VectorBase (or other better name)
Lets use boost::container::static_vector or boost::container::small_vector because dynamic allocations in the shape are painful:
<img width="867" alt="Screenshot 2024-10-22 at 3 13 52 PM" src="https://github.com/user-attachments/assets/0a719586-14e9-4626-8a3b-69fa52e6ec1e">

Reply:
* Lets rename mValue -> m_value
* vector_base -> VectorBase (or other better name)

Done

Looking into static_vector

Reply:
small_vector change makes sense, but should be done separately

Reply:
@dmakoviichuk-tt Yes, I wanted to use some kind of small_vector from my first day here :)
I agree with @ayerofieiev-tt, let's make the change separately. I think we'll have to go through the whole codebase eventually and cleanup vectors